tests/e2e/utils/payments.js

```diff
@@ -56,19 +56,24 @@ export async function fillCardDetails( page, card ) {
 			{ timeout: 30000 }
 		);
 		await cardNumberInput.type( card.number, { delay: 20 } );
+		await page.waitFor( 1000 );
 
 		const cardDateInput = await stripeFrame.waitForSelector(
-			'[name="exp-date"]'
+			'[name="exp-date"]',
+			{ timeout: 30000 }
 		);
 
 		await cardDateInput.type( card.expires.month + card.expires.year, {
 			delay: 20,
 		} );
+		await page.waitFor( 1000 );
```

Comment:
Not a huge fan of timeouts like these in tests tbh since we never really know whether they're useful, I have seen waiting for certain selectors or for the page to finish with network calls is better in terms of maintainability. WDYT?

Reply:
I agree 100%, and I tried multiple unsuccessful solutions before using this, which seems to help.

I ran a lot of times in a case where `tests/e2e/specs/wcpay/shopper/shopper-checkout-failures.spec.js` would have 1 or 2 tests fail because somehow the card number input didn't reflect exactly what we tried to fill in.
I'm not sure exactly what the problem is, something related to Puppeteer typing very fast (too fast?) and not waiting correctly for the next step...?

Take this run for instance: https://github.com/Automattic/woocommerce-payments/actions/runs/4487977959

The test `should throw an error that the card expiration date is in the past` should type the card number `4242424242424242` but you can see in the artifact's screenshots that the card number's last 4 digits are `0004`, and so instead of seeing the expected error, we see the "card number is invalid" error because obviously this card number is wrong. I'm not even sure how we end up with this number.

There's an issue on the Puppeteer repo: https://github.com/puppeteer/puppeteer/issues/1648

Which is closed, but people are still adding comments to complain about this issue, even on recent versions.

I'm open to suggestions if you have any though

Reply:
Fair, let's revisit in the future in case we're able to debug the flaky nature better.